### PR TITLE
Enchantment validation and nerf

### PR DESCRIPTION
**Pull request by uecasm**
# Changes proposed in this pull request:
- Adds some enchantment level validation to the enchanter datagen.
- Nerfs the enchanter so they don't produce greater-than-vanilla+1 enchantment levels any more.
    - ~~Were any of the greater-than-vanilla levels deliberate?  Many of those don't do anything, but a few do.~~
    - As noted below, one extra level is allowed above max where it actually does a thing.

Review please (can port)

Also included in the validation (but not [yet] fixed) is some future-proofing to log which enchants exist but are not produced by the enchanter (along with a means to hide the warnings if we decide that this is deliberate).  Currently these enchants are:

```
Max enchanter does not produce Impaling
Max enchanter does not produce Riptide
Max enchanter does not produce Channeling
Max enchanter does not produce Loyalty
Max enchanter does not produce Piercing
Max enchanter does not produce Thorns
Max enchanter does not produce Soul Speed
Max enchanter does not produce Swift Sneak
Max enchanter does not produce Luck of the Sea
Max enchanter does not produce Lure
```

Which, if any, of these are deliberate and which should we add in?

### CODE REVIEW

**Comment by Raycoms** [comment]
I wanted them to be able to produce vanilla+1 at max

**Comment by uecasm** [comment]
I've replaced the previous per-level table with a more algorithmic approach, which also removes the need for half of the validation.  I've changed it so that (where sensible) it will allow one extra level above the vanilla max.

Since the JSON diffs can be a bit hard to read, especially since some of the entries moved, here's the list of actual changes from pre-PR:

- Enchanter 1
    - No changes
- Enchanter 2
    - Remove Aqua Affinity II (pointless)
    - Remove Flame II (pointless)
- Enchanter 3
    - Remove Aqua Affinity II (pointless)
    - Remove Aqua Affinity III (pointless)
    - Remove Flame II (pointless)
    - Remove Flame III (pointless)
- Enchanter 4
    - Remove Aqua Affinity II (pointless)
    - Remove Aqua Affinity III (pointless)
    - Remove Aqua Affinity IV (pointless)
    - Remove Flame II (pointless)
    - Remove Flame III (pointless)
    - Remove Flame IV (pointless)
    - Remove Depth Strider IV (pointless)
    - Remove Fire Aspect IV (2 levels above max)
    - Remove Frost Walker IV (2 levels above max)
    - Remove Knockback IV (2 levels above max)
    - Remove Punch IV (2 levels above max)
    - Replace Infinity IV with Infinity I (pointless)
    - Add Protection IV (w=5) [mysteriously omitted]
- Enchanter 5
    - Remove Aqua Affinity III (pointless)
    - Remove Aqua Affinity IV (pointless)
    - Remove Aqua Affinity V (pointless)
    - Remove Flame III (pointless)
    - Remove Flame IV (pointless)
    - Remove Flame V (pointless)
    - Remove Depth Strider IV (pointless)
    - Remove Depth Strider V (pointless)
    - Remove Fire Aspect IV (2 levels above max)
    - Remove Fire Aspect V (3 levels above max)
    - Remove Frost Walker IV (2 levels above max)
    - Remove Frost Walker V (3 levels above max)
    - Remove Knockback IV (2 levels above max)
    - Remove Knockback V (3 levels above max)
    - Remove Punch IV (2 levels above max)
    - Remove Punch V (3 levels above max)
    - Remove Looting V (2 levels above max)
    - Remove Quick Charge V (2 levels above max)
    - Remove Respiration V (2 levels above max)
    - Remove Sweeping V (2 levels above max)
    - Remove Unbreaking V (2 levels above max)
    - Remove a second copy of Infinity I (effectively reducing w=2 to w=1)
    - Add Protection IV (w=5) [had III and V, but IV was strangely missing]

I _think_ the loot weights are all the same, other than as noted above, but I might have missed something.

The warnings about the other enchantments are still present, so we should decide what to do with them before merging too.

**Review by Raycoms — DISMISSED** [review]

**Comment by uecasm** [comment]
Tweaked the rules a bit:
* Fixed a bug where enchants with only one level were completely unavailable at enchanter 4-5; now available at reduced weight.
* Juggled some enchants around so that `VERY_RARE` and non-enchanting-table enchants are flagged as rare (weight 1 at all levels), with validation warning for future violations
* Removed extra bonus levels from enchants that already have 5 levels, so couldn't have had a bonus anyway.  Added validation warning for future violations.
* Added missing enchants; most of these are rares.

The changes from the previous list are as follows:
- Enchanter 1
  - Remove Frost Walker I
  - Add Piercing I [w=50]
  - Add Impaling I [w=1]
  - Add Riptide I [w=1]
- Enchanter 2
  - Remove Frost Walker II (keep Frost Walker I [w=1])
  - Add Piercing I / II [w=50/25]
  - Add Impaling I / II [w=1]
  - Add Riptide I / II [w=1]
  - Add Loyalty I [w=1]
  - Add Thorns I [w=1]
- Enchanter 3
  - Remove Frost Walker III (keep Frost Walker I / II [w=1])
  - Add Piercing I / II / III [w=50/25/15]
  - Add Impaling I / II / III [w=1]
  - Add Riptide I / II / III [w=1]
  - Add Loyalty I / II [w=1]
  - Add Thorns I / II [w=1]
  - Add Channeling I [w=1]
  - Add Swift Sneak I [w=1]
  - Add Soul Speed I [w=1]
- Enchanter 4
  - Keep Frost Walker I / II / III [w=1]
  - Change Infinity I [w=5] to [w=1]
  - Add Aqua Affinity I [w=5]
  - Add Flame I [w=5]
  - Add Piercing II / III / IV [w=25/15/5]
  - Add Impaling II / III / IV [w=1]
  - Add Riptide II / III / IV [w=1]
  - Add Loyalty I / II / III [w=1]
  - Add Thorns I / II [w=1]
  - Add Channeling I [w=1]
  - Add Swift Sneak I / II [w=1]
  - Add Soul Speed I / II [w=1]
- Enchanter 5
  - Keep Frost Walker II / III [w=1]
  - Change Infinity I [w=5] to [w=1]
  - Add Aqua Affinity I [w=5]
  - Add Flame I [w=5]
  - Add Piercing III / IV / V [w=15/5/1]
  - Add Impaling III / IV / V [w=1]
  - Add Riptide III / IV [w=1]
  - Add Loyalty II / III / IV [w=1]
  - Add Thorns II / III / IV [w=1]
  - Add Channeling I [w=1]
  - Add Swift Sneak I / II / III [w=1]
  - Add Soul Speed I / II / III [w=1]

**Review by Raycoms — APPROVED** [review]

**Pull request merged**
